src/threading_utils/parallel_for.h

```diff
@@ -53,8 +54,38 @@ class OMPException {
   }
 };
 
+inline int OmpGetThreadLimit() {
+  int limit = omp_get_thread_limit();
+  TREELITE_CHECK_GE(limit, 1) << "Invalid thread limit for OpenMP.";
+  return limit;
+}
+
 inline int MaxNumThread() {
-  return omp_get_max_threads();
+  return std::min(std::min(omp_get_num_procs(), omp_get_max_threads()), OmpGetThreadLimit());
+}
+
+/*!
+ * \brief Represent thread configuration, to be used with parallel loops.
+ */
+struct ThreadConfig {
+  int nthread;
```

Comment:
Could we make this a `std::optional<std::size_t>`? We're only using the signed integer for special value signaling, right?

Reply:
Let's use uint32_t actually. It doesn't need to be optional because we always first validate the user input.